File: src/models/currency.nim
import std/[times, options]
import norm/[model, pragmas]

type
  Currency* = ref object of Model
    code* {.unique.}: string
    name*: string
    name_bg*: string
    symbol*: string
    decimal_places*: int
    is_active*: bool
    is_base_currency*: bool
    bnb_code*: string
    created_at*: DateTime
    updated_at*: DateTime

proc newCurrency*(
  code = "",
  name = "",
  name_bg = "",
  symbol = "",
  decimal_places = 2,
  is_active = true,
  is_base_currency = false,
  bnb_code = ""
): Currency =
  Currency(
    code: code,
    name: name,
    name_bg: name_bg,
    symbol: symbol,
    decimal_places: decimal_places,
    is_active: is_active,
    is_base_currency: is_base_currency,
    bnb_code: bnb_code,
    created_at: now(),
    updated_at: now()
  )

File: baraba_ub/apps/baraba_umbrella/lib/baraba_umbrella/accounting/currency.ex
defmodule BarabaUmbrella.Accounting.Currency do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id

  schema "currencies" do
    field :code, :string
    field :name, :string
    field :name_bg, :string
    field :symbol, :string
    field :decimal_places, :integer, default: 2
    field :is_active, :boolean, default: true
    field :is_base_currency, :boolean, default: false
    field :bnb_code, :string

    timestamps()
  end

  @doc false
  def changeset(currency, attrs) do
    currency
    |> cast(attrs, [:code, :name, :name_bg, :symbol, :decimal_places, :is_active, :is_base_currency, :bnb_code])
    |> validate_required([:code, :name])
    |> unique_constraint(:code)
  end
end
